### PR TITLE
drivers: xilinx: Select PINCTRL

### PR DESCRIPTION
**Pull request by nordicjm**
drivers: xilinx: Select PINCTRL
    
    Selects PINCTRL in drivers

    boards: digilent: zybo: Do not select PINCTRL
    
    This should not be selected by boards

    boards: enclustra: mercury_xu: Do not select PINCTRL
    
    This should not be selected by boards

### CODE REVIEW

File: drivers/gpio/Kconfig.xlnx [hunk 1]
```diff
@@ -7,5 +7,6 @@ config GPIO_XLNX_AXI
 	bool "Xilinx AXI GPIO driver"
 	default y
 	depends on DT_HAS_XLNX_XPS_GPIO_1_00_A_ENABLED
+	select PINCTRL
```
Comment (henrikbrixandersen) [inline drivers/gpio/Kconfig.xlnx:10]: This driver does not use pinctrl (pins are routed in the FPGA design, there is no pin controller associated).

File: drivers/gpio/Kconfig.xlnx_ps [hunk 2]
```diff
@@ -11,5 +11,6 @@ config GPIO_XLNX_PS
 	default y
 	depends on DT_HAS_XLNX_PS_GPIO_ENABLED
 	depends on !QEMU_TARGET
+	select PINCTRL
```
Comment (henrikbrixandersen) [inline drivers/gpio/Kconfig.xlnx_ps:14]: This driver does not use pinctrl.

File: drivers/i2c/Kconfig.xilinx_axi [hunk 3]
```diff
@@ -6,6 +6,7 @@ config I2C_XILINX_AXI
 	default y
 	depends on DT_HAS_XLNX_XPS_IIC_2_00_A_ENABLED || DT_HAS_XLNX_XPS_IIC_2_1_ENABLED
 	select EVENTS
+	select PINCTRL
```
Comment (henrikbrixandersen) [inline drivers/i2c/Kconfig.xilinx_axi:9]: This driver does not use pinctrl.

File: drivers/serial/Kconfig.xlnx [hunk 4]
```diff
@@ -9,6 +9,7 @@ config UART_XLNX_PS
 	depends on DT_HAS_XLNX_XUARTPS_ENABLED
 	select SERIAL_HAS_DRIVER
 	select SERIAL_SUPPORT_INTERRUPT
+	select PINCTRL
```
Comment (henrikbrixandersen) [inline drivers/serial/Kconfig.xlnx:12]: Pinctrl is optional in this driver, as the pins can be routed in the FPGA.

File: drivers/serial/Kconfig.xlnx [hunk 5]
```diff
@@ -18,5 +19,6 @@ config UART_XLNX_UARTLITE
 	depends on DT_HAS_XLNX_XPS_UARTLITE_1_00_A_ENABLED
 	select SERIAL_HAS_DRIVER
 	select SERIAL_SUPPORT_INTERRUPT
+	select PINCTRL
```
Comment (henrikbrixandersen) [inline drivers/serial/Kconfig.xlnx:22]: This driver does not use pinctrl.

File: drivers/spi/Kconfig.xlnx [hunk 6]
```diff
@@ -8,5 +8,6 @@ config SPI_XLNX_AXI_QUADSPI
 	default y
 	depends on DT_HAS_XLNX_XPS_SPI_2_00_A_ENABLED
 	select EVENTS
+	select PINCTRL
```
Comment (henrikbrixandersen) [inline drivers/spi/Kconfig.xlnx:11]: This driver does not use pinctrl.

File: boards/digilent/zybo/zybo_defconfig [hunk 7]
```diff
@@ -4,7 +4,6 @@
 CONFIG_ARM_ARCH_TIMER=y
 CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=325000000
 
-CONFIG_PINCTRL=y
```
Comment (henrikbrixandersen) [inline boards/digilent/zybo/zybo_defconfig:7]: For FPGA-based designs like this, it actually is up to the board to select pinctrl if needed by the design. Some designs route every pin in the FPGA, som do it at runtime.

File: boards/enclustra/mercury_xu/mercury_xu_defconfig [hunk 8]
```diff
@@ -9,5 +9,3 @@ CONFIG_UART_CONSOLE=y
 
 # Enable timer
 CONFIG_SYS_CLOCK_TICKS_PER_SEC=100
-
-CONFIG_PINCTRL=y
```
Comment (henrikbrixandersen) [inline boards/enclustra/mercury_xu/mercury_xu_defconfig:13]: Same here.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/spi/Kconfig.xlnx (hunk 6 above)
Comment (nordicjm) [reply drivers/spi/Kconfig.xlnx]: Removed

File: drivers/serial/Kconfig.xlnx (hunk 5 above)
Comment (nordicjm) [reply drivers/serial/Kconfig.xlnx]: Removed

File: drivers/serial/Kconfig.xlnx (hunk 4 above)
Comment (nordicjm) [reply drivers/serial/Kconfig.xlnx]: Removed

File: drivers/i2c/Kconfig.xilinx_axi (hunk 3 above)
Comment (nordicjm) [reply drivers/i2c/Kconfig.xilinx_axi]: Removed

File: drivers/gpio/Kconfig.xlnx_ps (hunk 2 above)
Comment (nordicjm) [reply drivers/gpio/Kconfig.xlnx_ps]: Removed

File: drivers/gpio/Kconfig.xlnx (hunk 1 above)
Comment (nordicjm) [reply drivers/gpio/Kconfig.xlnx]: Removed

File: boards/digilent/zybo/zybo_defconfig (hunk 7 above)
Comment (henrikbrixandersen) [reply boards/digilent/zybo/zybo_defconfig:7]: This needs to stay as pinctrl is optional in `drivers/serial/uart_xlnx_ps.c`.

File: boards/enclustra/mercury_xu/mercury_xu_defconfig (hunk 8 above)
Comment (henrikbrixandersen) [reply boards/enclustra/mercury_xu/mercury_xu_defconfig:13]: This needs to stay as pinctrl is optional in `drivers/serial/uart_xlnx_ps.c`.

File: drivers/ethernet/Kconfig.xlnx_gem [hunk 9]
```diff
@@ -11,6 +11,7 @@ menuconfig ETH_XLNX_GEM
 	default y
 	depends on DT_HAS_XLNX_GEM_ENABLED
 	depends on !QEMU_TARGET || (QEMU_TARGET && NET_QEMU_ETHERNET)
+	select PINCTRL
```
Comment (henrikbrixandersen) [inline drivers/ethernet/Kconfig.xlnx_gem:14]: This driver does not use pinctrl either.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/ethernet/Kconfig.xlnx_gem (hunk 9 above)
Comment (nordicjm) [reply drivers/ethernet/Kconfig.xlnx_gem]: Removed

**Review by henrikbrixandersen — APPROVED** [review]
Thanks

**Review by ibirnbaum — APPROVED** [review]

**Comment by ibirnbaum** [comment]
Although this wasn't mentioned here: the Ethernet controller is another instance of a peripheral where the pins may either be routed to external I/O pins or into the programmable logic part of the chip, i.e. where PINCTRL wouldn't be universally applicable.

**Comment by ibirnbaum** [comment]
In theory, it should be possible to assign the MIO pins for the Ethernet controller's RGMII interface signals via PINCTRL, but the alternative of routing the GMII interface into the FPGA via EMIO likely can't be configured that way.

I agree that PINCTRL should be specified at the board level for the Xilinx SoCs, as, when using the Xilinx FSBL or u-boot SPL derived from the system design in Vivado, which is a pretty standard approach, everything relating to the I/O configuration has already been handled prior to the OS booting.

**Review by dcpleung — APPROVED** [review]

**Pull request merged**